myia/ir.py

```diff
@@ -1,7 +1,142 @@
-"""Intermediate representation definition."""
+"""Intermediate representation definition.
+
+Myia's main intermediate representation (IR) is a graph-based version of ANF.
+Each function definition (lambda) is defined as a graph, consisting of a series
+of function applications.
+
+A function can be applied to a node from another funtion's graph; this
+implicitly creates a nested function. Functions are first-class objects, so
+returning a nested function creates a closure.
+
+"""
+from enum import Enum, auto
+from typing import List, Union, Dict
+
+LiteralType = Union[str, int, float, 'PrimitiveOperation', 'Graph', None]
+
+
+class Graph:
+    """A function graph.
+
+    Attributes:
+        output: The `Node` whose value will be returned by this function.
+
+    """
+
+    def __init__(self, output: 'Node') -> None:
+        """Construct a graph."""
+        self.output = output
+        self.debug: Dict = {}
 
 
 class Node:
-    """A node in the graph-based ANF IR."""
+    """A node in the graph-based ANF IR.
+
+    There are three types of nodes: Values, which are the result of function
+    applications; parameters; and constants such as numbers and functions.
+
+    Attributes:
+        inputs: If the node is a value, the first node input is the function to
+            apply, followed by the arguments.
+        value: The value of this node, if it is a constant.
+        graph: The function definition graph that this node belongs to for
+            values and parameters.
+        debug: A dictionary with debug information about this node e.g. a
+            human-readable name and the Python source code.
+
+    """
+
+    def __init__(self) -> None:
+        """Construct a node."""
+        self.inputs: List[Node] = []
+        self.value: LiteralType = None
+        self.graph: Graph = None
+        self.debug: Dict = {}
+
+
+class Value(Node):
+    """A value.
+
+    A value is the result of function application.
+
+    """
+
+    def __init__(self, inputs: List[Node], graph: 'Graph') -> None:
+        """Construct a value."""
+        if len(inputs) < 1:
+            raise ValueError
+        super().__init__()
+        self.inputs = inputs
+        self.graph = graph
+
+
+class Parameter(Node):
+    """A parameter to a function.
+
+    Parameters are leaf nodes, since they are not the result of a function
+    application, and they have no value. They are entirely defined by the graph
+    they belong to.
+
+    """
+
+    def __init__(self, graph: Graph) -> None:
+        """Construct the parameter."""
+        super().__init__()
+        self.graph = graph
+
+
+class Constant(Node):
+    """A constant node.
+
+    A constant is a node which is not the result of a function application. In
+    the graph it is a leaf node. It has no inputs, and instead is defined
+    entirely by its value. Unlike parameters and values, constants do not
+    belong to any particular function graph.
+
```

Comment:
I would consider letting constants belong to graphs, because:

* I see no particular advantage in merging constants (maybe I'm missing something).
* Having a separate node for each occurrence of a constant means we can track the provenance of each separately. If, say, you have a function, you apply the AD transform, then a bunch of optimizations, and you're trying to debug where the hell some zero comes from, it's nice to be able to see it's the zero on line 38 and not the zero on line 47. I've definitely had this kind of issue before.

Reply:
The advantage of "merging" constants (CSE) propagates, in the sense that if the `0` in `(x + 0) + (x + 0)` becomes the same, it then becomes trivial to determine that the two terms for `x + 1` are also the same, etc. This allows you to do full-fledged CSE at construction time, which sea-of-nodes used as a big selling point.

I can see the advantage of debugging, but I don't think you need to assign nodes to graphs for that. You can simply create two constants with the same value, which will contain different debugging information.

Reply:
Yeah, and if I think about it, if we want to track multiple instances of constants, we may want to track multiple instances of every other node (for each use) which defeats the purpose.

So I think the right thing to do here would be to be able to attach debug info to *edges*, but I should elaborate on this elsewhere.